pkg/webhook/resources/virtualmachinebackup/validator.go

```diff
@@ -177,3 +190,66 @@ func (v *virtualMachineBackupValidator) Delete(_ *types.Request, obj runtime.Obj
 	}
 	return nil
 }
+
+func (v *virtualMachineBackupValidator) checkTotalSnapshotSize(vm *kubevirtv1.VirtualMachine) error {
+	resourceQuota, err := v.resourceQuotaCache.Get(vm.Namespace, util.DefaultResourceQuotaName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return werror.NewInternalError(fmt.Sprintf("failed to get resource quota %s/%s, err: %s", vm.Namespace, util.DefaultResourceQuotaName, err))
+	}
+
+	if resourceQuota.Spec.SnapshotLimit.VMTotalSnapshotSizeQuota != nil {
+		if err := webhookutil.CheckTotalSnapshotSizeOnVM(v.pvcCache, v.engineCache, vm, resourceQuota.Spec.SnapshotLimit.VMTotalSnapshotSizeQuota[vm.Name]); err != nil {
+			return err
+		}
+	}
+
+	if err := webhookutil.CheckTotalSnapshotSizeOnNamespace(v.pvcCache, v.engineCache, vm.Namespace, resourceQuota.Spec.SnapshotLimit.NamespaceTotalSnapshotSizeQuota); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (v *virtualMachineBackupValidator) checkTotalSnapshotSizeOnVM(vm *kubevirtv1.VirtualMachine, totalSnapshotSizeQuota int64) error {
+	if totalSnapshotSizeQuota == 0 {
+		return nil
+	}
+
+	var totalSnapshotUsage int64
+	for _, volume := range vm.Spec.Template.Spec.Volumes {
+		if volume.PersistentVolumeClaim == nil {
+			continue
+		}
+
+		pvcNamespace := vm.Namespace
+		pvcName := volume.PersistentVolumeClaim.ClaimName
+
+		pvc, err := v.pvcCache.Get(pvcNamespace, pvcName)
+		if err != nil {
+			return werror.NewInternalError(fmt.Sprintf("failed to get PVC %s/%s, err: %s", pvcNamespace, pvcName, err))
+		}
+
+		engine, err := webhookutil.GetLHEngine(v.engineCache, pvc.Spec.VolumeName)
```

Comment:
We should check if a VM volume is a Longhorn volume first. The implementation counts Longhorn volume only and it's possible to attach 3rd party storage volume to a VM. Can you also check if the following code works with Longhorn v2 volumes? Thanks.

Same with the volumesnapshot validator, not 100% sure if the code works on non-Longhorn volumes.

Reply:
Good catch. I update it to check whether PVC is from LH first. For now, this feature can only work with LH. We may need to have an adapter for different driver.